nbcelltests/tests/test_test.py

```diff
@@ -5,9 +5,223 @@
 # This file is part of the nbcelltests library, distributed under the terms of
 # the Apache License 2.0.  The full license can be found in the LICENSE file.
 #
-# for Coverage
+import tempfile
+import os
+import sys
+import unittest
 
+from nbcelltests.test import run
 
-class TestTests:
-    def test_run(self):
+# TODO: we should generate the notebooks rather than having them as files
+# (same for lint ones)
+CUMULATIVE_RUN = os.path.join(os.path.dirname(__file__), '_cumulative_run.ipynb')
+CELL_ERROR = os.path.join(os.path.dirname(__file__), '_cell_error.ipynb')
+TEST_ERROR = os.path.join(os.path.dirname(__file__), '_test_error.ipynb')
+TEST_FAIL = os.path.join(os.path.dirname(__file__), '_test_fail.ipynb')
+
+# Hack. We want to test expected behavior in distributed situation,
+# which we are doing via pytest --forked.
+FORKED = '--forked' in sys.argv
+
+
+def _assert_x_undefined(t):
+    t.run_test("""
+    try:
+        x
+    except NameError:
         pass
+    else:
+        raise Exception('x was already defined')
+    """)
+
+# TODO: This test file's manual use of unittest is brittle
+
+
+def _import_from_path(pypath, f):
+    import importlib.util
+    spec = importlib.util.spec_from_file_location(pypath, f)
+    mod = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(mod)
+    return mod
+
+
+class _TestCellTests(unittest.TestCase):
+    # abstract :)
+
+    @classmethod
+    def setUpClass(cls):
+        """
+        Generate test file from notebook, then import it, and make the
+        resulting module available as "generated_tests" class attribute.
+        """
+        tf = tempfile.NamedTemporaryFile(mode='w', suffix='.py', delete=False, encoding='utf8')
+        tf_name = tf.name
+        try:
+            cls.generated_tests = _import_from_path("nbcelltests.tests.%s.%s" % (__name__, cls.__name__), run(cls.NBNAME, filename=tf_name))
+            tf.close()
+        finally:
+            os.remove(tf_name)
+
+
+class TestCumulativeRun(_TestCellTests):
+
+    NBNAME = CUMULATIVE_RUN
+
+    def test_state(self):
+        """
+                    cell    test   state
+        0:           -       -       -
+
+        1: either   x=0      -      x=0   (expected)
+               or    -       -      x>0   (bad; x was previously defined in kernel)
```

Comment:
What would trigger the `or` condition of this either/or branching? Can you add a short comment on the top describing how to read the table?

Reply:
It's tricky - if you can suggest something better I'll be grateful. 

If `x` is already defined, nothing is done. If `x` is not defined, it's set to 0. This is how I caught the case that the tests were running all the cells multiple times in the same kernel.

Reply:
> If x is already defined, nothing is done. If x is not defined, it's set to 0. 

How do you read this out of the table though? 😕

Reply:
I said it was tricky! :) I'll think about it - I'm sure we can do better.